tracing/opentracing/id_extract.go

```diff
@@ -12,39 +9,56 @@ import (
 )
 
 const (
-	TagTraceId = "trace.traceid"
-	TagSpanId  = "trace.spanid"
+	TagTraceId           = "trace.traceid"
+	TagSpanId            = "trace.spanid"
+	TagSampled           = "trace.sampled"
+	jaegerNotSampledFlag = "0"
 )
 
-// hackyInjectOpentracingIdsToTags writes the given context to the ctxtags.
+// injectOpentracingIdsToTags writes trace data to ctxtags.
 // This is done in an incredibly hacky way, because the public-facing interface of opentracing doesn't give access to
 // the TraceId and SpanId of the SpanContext. Only the Tracer's Inject/Extract methods know what these are.
 // Most tracers have them encoded as keys with 'traceid' and 'spanid':
 // https://github.com/openzipkin/zipkin-go-opentracing/blob/594640b9ef7e5c994e8d9499359d693c032d738c/propagation_ot.go#L29
 // https://github.com/opentracing/basictracer-go/blob/1b32af207119a14b1b231d451df3ed04a72efebf/propagation_ot.go#L26
 // Jaeger from Uber use one-key schema with next format '{trace-id}:{span-id}:{parent-span-id}:{flags}'
 // https://www.jaegertracing.io/docs/client-libraries/#trace-span-identity
-func hackyInjectOpentracingIdsToTags(span opentracing.Span, tags grpc_ctxtags.Tags) {
-	if err := span.Tracer().Inject(span.Context(), opentracing.HTTPHeaders, &hackyTagsCarrier{tags}); err != nil {
-		grpclog.Printf("grpc_opentracing: failed extracting trace info into ctx %v", err)
+func injectOpentracingIdsToTags(span opentracing.Span, tags grpc_ctxtags.Tags) {
+	if err := span.Tracer().Inject(span.Context(), opentracing.HTTPHeaders, &tagsCarrier{tags}); err != nil {
+		grpclog.Infof("grpc_opentracing: failed extracting trace info into ctx %v", err)
 	}
 }
 
-// hackyTagsCarrier is a really hacky way of
-type hackyTagsCarrier struct {
+// tagsCarrier is a really hacky way of
+type tagsCarrier struct {
 	grpc_ctxtags.Tags
 }
 
-func (t *hackyTagsCarrier) Set(key, val string) {
-	if strings.Contains(key, "traceid") || strings.Contains(strings.ToLower(key), "traceid") {
+func (t *tagsCarrier) Set(key, val string) {
+	key = strings.ToLower(key)
+	if strings.Contains(key, "traceid") {
 		t.Tags.Set(TagTraceId, val) // this will most likely be base-16 (hex) encoded
-	} else if (strings.Contains(key, "spanid") && !strings.Contains(key, "parent")) || (strings.Contains(strings.ToLower(key), "spanid") && !strings.Contains(strings.ToLower(key), "parent")) {
+	}
+
+	if strings.Contains(key, "spanid") && !strings.Contains(strings.ToLower(key), "parent") {
 		t.Tags.Set(TagSpanId, val) // this will most likely be base-16 (hex) encoded
-	} else if key == "uber-trace-id" {
+	}
+
+	if strings.Contains(key, "sampled") {
+		t.Tags.Set(TagSampled, val)
```

Comment:
How about enforcing that the val is either `true` or `false`, and ignoring it if it isn't?

Reply:
We could do that however both OpenZipkin and OpenTracing do `strconv.FormatBool(sc.Sampled)` to set the value

Reply:
I think I would prefer to pass thru what is in the sampled field from the OT implementation if it is not true/false it would be valuable to have this visualised.

Reply:
@devnev is this change blocking a merge?